### ISSUE TITLE
`no-undefined-types` treat normal variables as `undefined` in function scope

### ISSUE THREAD

**Issue by hecticme** [gajus/eslint-plugin-jsdoc#1342]
Reproduction: https://stackblitz.com/edit/vitejs-vite-d1gqkb

## Expected behavior

I expected that this code snippet does not cause any lint error:

```js
function retrieveHexCode(colorName) {
  const hashmap = {
    white: '#fff',
  };

  // 👇️ Lint error: The type 'hashmap' is undefined.
  /** @typedef {keyof typeof hashmap} KeyOfHashmap */
  // It seems like normal variables are not recognized.
  /** @typedef {typeof hashmap} TypeOfHashmap */

  const colorHexCode = hashmap[/** @type {KeyOfHashmap} */ (colorName)];

  return colorHexCode ?? '#000';
}

const hashmapTwo = {
  white: '#fff',
};

// 👇️ However, the error does not occur outside of function scope.
/** @typedef {typeof hashmapTwo} TypeOfHashmapTwo */
```

## Actual behavior

In the above snippet, `hashmap` is reported as `undefined` by `jsdoc/no-undefined-types`. This only happens inside the function scope, please take a look at the reproduction playground for more details. Intellisense was still working properly.

## ESLint Config

```js
import jsdoc from 'eslint-plugin-jsdoc';

export default [
  {
    files: ['**/*.js'],
    plugins: {
      jsdoc,
    },
    rules: {
      'jsdoc/no-undefined-types': [
        'error',
        {
          definedTypes: [],
          disableReporting: false,
          markVariablesAsUsed: true,
        },
      ],
    },
  },
];
```

## ESLint sample

See → https://stackblitz.com/edit/vitejs-vite-d1gqkb or check the code in "Expected behavior".

## Environment

- Node version: v18.20.3
- ESLint version: v9.15.0
- `eslint-plugin-jsdoc` version: 50.5.0

**Comment by RobertAKARobin** [comment]
I'm experiencing similar unexpected behavior with `typeof`.

**Comment by sebastian-fredriksson-bernholtz** [comment]
I think this is described in https://github.com/gajus/eslint-plugin-jsdoc/issues/178. That was closed as completed because typescript supposedly supports this now<sup>[1](https://github.com/gajus/eslint-plugin-jsdoc/issues/178#issuecomment-1566208281),[2](https://github.com/gajus/eslint-plugin-jsdoc/pull/1098#issue-1727984241)</sup>, but I haven't been able to figure out how to get typescript to flag invalid references in jsdoc. 

If anyone figures it out, would love to know how.

**Comment by hecticme** [comment]
Thanks Bernholtz :). I didn't see `#178` when I first searched for related issue. Probably because I forgot to remove the `is:open` filter.

Going to close this issue as "not planned" in a few days if no one else is interested.

**Comment by sebastian-fredriksson-bernholtz** [comment]
@hecticme This was probably obvious to you since your code is `js`, but mine is `ts`, so I only now figured out that what was being referred to was that it's recommended to use typescript to typecheck your js files to catch your issue.

I think it might still be worth **leaving your issue open** however, because

1. The rule should probably be marked deprecated and include an explanation that it's recommended to use typescript to typecheck your `js` files, instead of using this rule. 
Unless there is another open issue tracking that - or we create one, or reopen [#178](https://github.com/gajus/eslint-plugin-jsdoc/issues/178) - that you can close this in preference of, I think this isn't completely resolved. Although the issue would be one of lacking documentation instead.
2. I wouldn't actually consider the rule deprecated, because it still provides functionality that typescript does not (see [#178 (comment)](https://github.com/gajus/eslint-plugin-jsdoc/issues/178#issuecomment-2726894178)), so the underlying issue of not detecting local variables is still an issue, and unless that gets tracked in another issue, might as well keep this issue open.

**Comment by hecticme** [comment]
Sure thing. I will let the maintainer decide whether to close it or not.

**Comment by brettz9** [comment]
Fixed by #1363
